src/workerd/api/queue.c++

```diff
@@ -1,36 +1,114 @@
 #include "queue.h"
+#include "src/workerd/jsg/jsg.h"
 
 #include <workerd/jsg/ser.h>
 #include <workerd/api/global-scope.h>
 #include <kj/encoding.h>
 
 namespace workerd::api {
 
-kj::Promise<void> WorkerQueue::send(
-    v8::Local<v8::Value> body, jsg::Optional<SendOptions> options, v8::Isolate* isolate) {
-  auto& context = IoContext::current();
+kj::StringPtr validateContentType(kj::StringPtr contentType) {
+  if (contentType == IncomingQueueMessage::ContentType::TEXT) {
+    return IncomingQueueMessage::ContentType::TEXT;
+  }
+  if (contentType == IncomingQueueMessage::ContentType::OCTET_STREAM) {
+    return IncomingQueueMessage::ContentType::OCTET_STREAM;
+  }
+  if (contentType == IncomingQueueMessage::ContentType::JSON) {
+    return IncomingQueueMessage::ContentType::JSON;
+  }
+  if (contentType == IncomingQueueMessage::ContentType::V8) {
+    return IncomingQueueMessage::ContentType::V8;
+  }
 
-  JSG_REQUIRE(!body->IsUndefined(), TypeError, "Message body cannot be undefined");
+  JSG_FAIL_REQUIRE(TypeError, kj::str("Unsupported queue message content type: ", contentType));
+}
 
+kj::Array<kj::byte> serializeV8(jsg::Lock& js, v8::Local<v8::Value> body) {
   // Use a specific serialization version to avoid sending messages using a new version before all
   // runtimes at the edge know how to read it.
-  jsg::Serializer serializer(isolate, jsg::Serializer::Options {
+  jsg::Serializer serializer(js.v8Isolate, jsg::Serializer::Options {
     .version = 15,
     .omitHeader = false,
   });
   serializer.write(body);
-  kj::Array<kj::byte> serialized = serializer.release().data;
+  return serializer.release().data;
+}
 
-  auto client = context.getHttpClient(subrequestChannel, true, nullptr, "queue_send"_kj);
+kj::Array<kj::byte> serialize(jsg::Lock& js, v8::Local<v8::Value> body, kj::StringPtr contentType) {
+  if (contentType == IncomingQueueMessage::ContentType::TEXT) {
+    JSG_REQUIRE(
+      body->IsString(),
+      TypeError,
+      kj::str(
+        "Content Type \"",
+        IncomingQueueMessage::ContentType::TEXT,
+        "\" requires a value of type string, but received: ",
+        body->TypeOf(js.v8Isolate)
+      )
+    );
+
+    kj::String s = js.toString(body);
+    return kj::heapArray(s.asBytes());
+  }
+  if (contentType == IncomingQueueMessage::ContentType::OCTET_STREAM) {
+    JSG_REQUIRE(
+      body->IsArrayBuffer() || body->IsUint8Array(),
```

Comment:
Why limit to `Uint8Array`? Any `body->IsArrayBufferView()` should work here also.

Reply:
Good point, I just wanted to start simple, but I think this should probably support the same full set of array-like types supported by R2's `put()`: https://developers.cloudflare.com/r2/api/workers/workers-api-reference/#bucket-method-definitions

Is checking `body->IsArrayBufferView()` sufficient? Or do I need to check a bunch of various types.

Reply:
`IsArrayBufferView()` should be enough.

Reply:
Done, checking `isArrayBufferView()` only now.